src/nupic/bindings/algorithms.i

```diff
@@ -1507,6 +1509,170 @@ inline PyObject* generate2DGaussianSample(nupic::UInt32 nrows, nupic::UInt32 nco
+        actValue = classification["actValue"]
+        category = False
+      elif classification["actValue"] is None:
+        # Use the sentinel value so we know if it gets used in actualValues
+        # returned.
+        actValue = noneSentinel
+        # Turn learning off this step.
+        learn = False
+        category = False
+        # This does not get used when learning is disabled anyway.
+        classification["bucketIdx"] = 0
+        isNone = True
+      else:
+        actValue = int(classification["bucketIdx"])
+        category = True
+
+      result = self.convertedCompute(
+          recordNum, patternNZ, int(classification["bucketIdx"]),
+          actValue, category, learn, infer)
+
+      if isNone:
+        for i, v in enumerate(result["actualValues"]):
+          if v - noneSentinel < 0.00001:
+            result["actualValues"][i] = None
+      arrayResult = dict((k, numpy.array(v)) if k != "actualValues" else (k, v)
+                         for k, v in result.iteritems())
+
+      if self.valueToCategory or isinstance(classification["actValue"], basestring):
+        # Convert the bucketIdx back to the original value.
+        for i in xrange(len(arrayResult["actualValues"])):
+          if arrayResult["actualValues"][i] is not None:
+            arrayResult["actualValues"][i] = self.valueToCategory.get(int(
+                arrayResult["actualValues"][i]), classification["actValue"])
+
+        self.valueToCategory[actValue] = classification["actValue"]
+
+      return arrayResult
+
+    def __getstate__(self):
+      # Save the local attributes but override the C++ classifier with the
+      # string representation.
+      d = dict(self.__dict__)
+      d["this"] = self.getCState()
+      return d
+
+    def __setstate__(self, state):
+      # Create an empty C++ classifier and populate it from the serialized
+      # string.
+      self.this = _ALGORITHMS.new_SDRClassifier()
+      if isinstance(state, str):
+        self.loadFromString(state)
+        self.valueToCategory = {}
+      else:
+        assert state["version"] == 0
+        self.loadFromString(state["this"])
+        # Use the rest of the state to set local Python attributes.
+        del state["this"]
+        self.__dict__.update(state)
+
+    @classmethod
+    def read(cls, proto):
+      instance = cls()
+      instance.convertedRead(proto)
+      return instance
+  %}
+
+  void loadFromString(const std::string& inString)
+  {
+    std::istringstream inStream(inString);
+    self->load(inStream);
+  }
+
+  PyObject* getCState()
+  {
+    SharedPythonOStream py_s(self->persistentSize());
+    std::ostream& s = py_s.getStream();
+    // TODO: Consider writing floats as binary instead.
+    s.flags(ios::scientific);
+    s.precision(numeric_limits<double>::digits10 + 1);
+    self->save(s);
+    return py_s.close();
+  }
+
+  PyObject* convertedCompute(UInt recordNum, const vector<UInt>& patternNZ,
```

Comment:
Several of these method names don't sound right. Our function/method names should be an action. `convertedRead` is another one in this category.

CC @scottpurdy